### PR TITLE
Budget App Project

### PR DESCRIPTION
**Pull request by faisalakb**
In this project apply the following features: 

Sign up and log in pages

    The user should be able to register in the app with full name, email and password (all mandatory).
    The user can log into the app using email and password.
    If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

Home page (categories page)

    When the user logs in, they are presented with the categories page.
    For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
    When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
    There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

Transactions page

    For a given category, the list of transactions is presented, ordered by the most recent.
    At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
    There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
    When the user clicks on the "Back" button (<), the user navigates to the home page.

"Add a new category" page

    The user fills out a form to create a new category, indicating their name and icon (both mandatory).
    The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
    When the user clicks on the "Back" button (<), the user navigates to the home page.

"Add a new transaction" page

    The user fills out a form to create a new transaction with:
        name (mandatory)
        amount (mandatory)
        categories (mandatory at least one)
    The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
    When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

### CODE REVIEW

File: app/views/devise/sessions/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  <div class="field">
+    <%= f.password_field :password, autocomplete: "current-password", placeholder: "Password" %>
+  </div>
+
+  <% if devise_mapping.rememberable? %>
+    <div class="field">
+      <%= f.check_box :remember_me %>
+      <%= f.label :remember_me %>
+    </div>
+  <% end %>
```
Comment (Ibizugbe) [inline app/views/devise/sessions/new.html.erb]: Nice job so far👍🏼. As a requirement, we should be welcomed with the splash screen that shows the name of your app and an option to log in and sign in as presented in the design😉.

splash screen design
image


Also, we are required to ensure we stick to the design, hence kindly ensure your login and sign up page matches the design👍🏼.

Login
image

Signup
image




Take a look at the spacing and positioning of each element and ensure they match the given design👍🏼.

File: app/views/categories/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  <br/>
+<div class="d-flex flex-column align-item-center" >
+<% @categories.each do |category| %>
+<%= link_to  category_transactions_path(category), style: "text-decoration: none; color: inherit;"  do %>
+  <div class="category">
+  <i class="<%= category.icon %> custom-icon fa-2x" ></i>
+  <h2 class="catName"><%= category.name %></h2>
+    
+  <p class="catAmount">Total Amount: <%= number_to_currency(category.total_amount) %></p>
+
```
Comment (Ibizugbe) [inline app/views/categories/index.html.erb:27]: Kindly do well to ensure your transaction and category page matches the given design as this is also a requirement in this milestone😉😉

your category page

image

Your transaction page
image


design
image


Kindly take a look at the positioning of each element, the border around them, and spacing, and ensure everything matches the design😉.

File: spec/controllers/categories_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+      # Perform a GET request to the show action
+      get :show, params: { id: category.id }
+
+      # Expect the response to redirect to the sign-in page
+      expect(response).to redirect_to(new_user_session_path)
+    end
+  end
+
+  # Add similar tests for GET #new and POST #create actions
```
Comment (Ibizugbe) [inline spec/controllers/categories_controller_spec.rb:24]: Nice job so far👍🏼. As you have instructed on your readme, I get this error when I try to run your test😉.
 Please, Lets do well to ensure all tests are passing before we proceed for a review as this is a requirement👍🏼
image

File: app/views/transactions/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  <br/>
+
+<%= form_with(model: [@category, @transaction], url: category_transactions_path(@category), method: :post) do |form| %>
+  <% if @transaction.errors.any? %>
+    <div id="error_explanation">
+      <h2><%= pluralize(@transaction.errors.count, "error") %> prohibited this transaction from being saved:</h2>
+      <ul>
+        <% @transaction.errors.full_messages.each do |message| %>
+          <li><%= message %></li>
+        <% end %>
```
Comment (Ibizugbe) [inline app/views/transactions/new.html.erb]: As a requirement, kindly implement a dropdown button for the categories the transaction will be associated to😉.

image

yours currently
image

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @faisalakb 👋🏼👋🏼, 

request change


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights✅✅✅

- [x] Form validations are properly implemented👍🏼.
- [x] routes are properly protected👍🏼.
- [x] The back button is properly implemented👍🏼.
- [x] User authentication works fine👍🏼.
- [x] Video presentation and documentation look great👍🏼.


## Required Changes ♻️

- [ ] For every implementation, try to ensure they reflect in the live version too😉.

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/controllers/categories_controller_spec.rb (hunk 3 above)
Comment (faisalakb) [reply spec/controllers/categories_controller_spec.rb:24]: @Ibizugbe  Please note that you have to use : **rspec spec** to run the tests. even i mention it in the ReadMe file as well

**Comment by faisalakb** [comment]
I resolve all the above comments mention above.

**Review by Reem-lab — APPROVED** [review]
# Hi @faisalakb,  👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights

- [ ]  Descriptive PR⭐
- [ ]  No Linters error 🤩
- [ ] Good job working on controllers methods💪🏻
- [ ] Good job working on models and validations 🙏🏻
- [ ] Great job working on the views🔥🔥
- [ ] Great design ⭐
- [ ] Huge efforts on the tests 🔥🔥
- [ ] Kudosn for working on the deployment ✅
- [ ] Great fir presenting the video 💪🏻
- [ ] Professional README 🚀
- [ ] Authentication Setup with devise well 🔥
- [ ] Good job working on the requested changes ⭐🙏🏻



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**